pkg/clusteragent/admission/controllers/webhook/controller.go

```diff
@@ -0,0 +1,260 @@
+	controller := &Controller{
+		clientSet:      client,
+		config:         config,
+		secretsLister:  secretInformer.Lister(),
+		secretsSynced:  secretInformer.Informer().HasSynced,
+		webhooksLister: webhookInformer.Lister(),
+		webhooksSynced: webhookInformer.Informer().HasSynced,
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "webhooks"),
+		isLeaderFunc:   isLeaderFunc,
+	}
+	secretInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.handleSecret,
+		UpdateFunc: controller.handleSecretUpdate,
+		DeleteFunc: controller.handleSecret,
+	})
+	webhookInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    controller.handleWebhook,
+		UpdateFunc: controller.handleWebhookUpdate,
+		DeleteFunc: controller.handleWebhook,
+	})
+	return controller
+}
+
+// Run starts the controller to process Secret and Webhook
+// events after sync'ing the informer's cache.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	log.Infof("Starting webhook controller for secret %s/%s and webhook %s", c.config.GetSecretNs(), c.config.GetSecretName(), c.config.GetName())
+	defer log.Infof("Stopping webhook controller for secret %s/%s and webhook %s", c.config.GetSecretNs(), c.config.GetSecretName(), c.config.GetName())
+
+	if ok := cache.WaitForCacheSync(stopCh, c.secretsSynced, c.webhooksSynced); !ok {
+		return
+	}
+
+	go wait.Until(c.run, time.Second, stopCh)
+
+	// Trigger a reconciliation to create the Webhook if it doesn't exist
+	c.queue.Add(c.config.GetName())
+
+	<-stopCh
+}
+
+// handleSecret enqueues the targeted Secret object when an event occurs.
+// It can be a callback function for deletion and addition events.
+func (c *Controller) handleSecret(obj interface{}) {
+	if !c.isLeaderFunc() {
+		return
+	}
+	if object, ok := obj.(metav1.Object); ok {
+		if object.GetNamespace() == c.config.GetSecretNs() && object.GetName() == c.config.GetSecretName() {
+			c.enqueue(object)
+		}
+	}
+}
+
+// handleSecretUpdate handles the new Secret reported in update events.
+// It can be a callback function for update events.
+func (c *Controller) handleSecretUpdate(oldObj, newObj interface{}) {
+	if !c.isLeaderFunc() {
+		return
+	}
+	newSecret, ok := newObj.(*corev1.Secret)
+	if !ok {
+		log.Debugf("Expected Secret object, got: %v", newObj)
+		return
+	}
+	oldSecret, ok := oldObj.(*corev1.Secret)
+	if !ok {
+		log.Debugf("Expected Secret object, got: %v", oldObj)
+		return
+	}
+	if newSecret.ResourceVersion == oldSecret.ResourceVersion {
+		return
+	}
+	c.handleSecret(newObj)
+}
+
+// handleWebhook enqueues the targeted Webhook object when an event occurs.
+// It can be a callback function for deletion and addition events.
+func (c *Controller) handleWebhook(obj interface{}) {
+	if !c.isLeaderFunc() {
+		return
+	}
+	if object, ok := obj.(metav1.Object); ok {
+		if object.GetName() == c.config.GetName() {
+			c.enqueue(object)
```

Comment:
I think we are using the same `c.queue` for both object kind: `Secret` and `Webhook` is it expected?
should we use 2 differents queue?

Reply:
yes I'm aware of that, the reconcile function is generic and simple enough to handle both secret and webhook events. I don't see why we should separate the queues as it will require adding more (and somehow duplicated) code. 
Happy to reconsider using different queues if you think the current reconcile logic is not clear enough or if there is another good reason to use separate queues